rust/lance-linalg/src/kmeans.rs

```diff
@@ -167,6 +170,219 @@ fn hist_stddev(k: usize, membership: &[Option<u32>]) -> f32 {
         .sqrt()
 }
 
+trait KMeansAlgo<T: Num> {
+    /// Recompute the membership of each vector.
+    ///
+    /// Parameters:
+    ///
+    /// - *data*: a `N * dimension` floating array. Not necessarily normalized.
+    ///
+    fn compute_membership_and_loss(
+        centroids: &[T],
+        data: &[T],
+        dimension: usize,
+        distance_type: DistanceType,
+    ) -> (Vec<Option<u32>>, f64);
+
+    /// Construct a new KMeans model.
+    fn to_kmeans(
+        data: &[T],
+        dimension: usize,
+        k: usize,
+        membership: &[Option<u32>],
+        distance_type: DistanceType,
+    ) -> KMeans;
+}
+
+struct KMeansAlgoFloat<T: ArrowNumericType>
+where
+    T::Native: Float + Num,
+{
+    phantom_data: std::marker::PhantomData<T>,
+}
+
+impl<T: ArrowNumericType> KMeansAlgo<T::Native> for KMeansAlgoFloat<T>
+where
+    T::Native: Float + Dot + L2 + DivAssign + AddAssign + FromPrimitive + Sync,
+{
+    fn compute_membership_and_loss(
+        centroids: &[T::Native],
+        data: &[T::Native],
+        dimension: usize,
+        distance_type: DistanceType,
+    ) -> (Vec<Option<u32>>, f64) {
+        let cluster_and_dists = match distance_type {
+            DistanceType::L2 => data
+                .par_chunks(dimension)
+                .map(|vec| argmin_value(l2_distance_batch(vec, centroids, dimension)))
+                .collect::<Vec<_>>(),
+            DistanceType::Dot => data
+                .par_chunks(dimension)
+                .map(|vec| argmin_value(dot_distance_batch(vec, centroids, dimension)))
+                .collect::<Vec<_>>(),
+            _ => {
+                panic!(
+                    "KMeans::find_partitions: {} is not supported",
+                    distance_type
+                );
+            }
+        };
+        (
+            cluster_and_dists
+                .par_iter()
+                .map(|cd| cd.map(|(c, _)| c))
+                .collect::<Vec<_>>(),
+            cluster_and_dists
+                .par_iter()
```

Comment:
can we get improvement from this?
- the list is short (i would assume it won't be over 10K)
- single thread would be more cache-friendly

Reply:
or `par_iter` will detect it's short and actually does this with only single thread?